src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,767 @@
+    {
+        int numInnerNodes = leafBlockStartValues.size();
+        byte[] splitValues = new byte[(1 + numInnerNodes) * bytesPerValue];
+        int depth = recurseBalanceTree(1, 0, numInnerNodes, 1, splitValues, leafBlockStartValues);
+        long[] leafBlockFPs = leafBlockFilePointer.stream().mapToLong(l -> l).toArray();
+        byte[] packedIndex = packIndex(leafBlockFPs, splitValues);
+
+        out.writeVInt(countPerLeaf);
+        out.writeVInt(bytesPerValue);
+
+        out.writeVInt(leafBlockFPs.length);
+        out.writeVInt(Math.min(depth, leafBlockFPs.length));
+
+        out.writeBytes(minPackedValue, 0, bytesPerValue);
+        out.writeBytes(maxPackedValue, 0, bytesPerValue);
+
+        out.writeVLong(valueCount);
+
+        out.writeVInt(packedIndex.length);
+        out.writeBytes(packedIndex, 0, packedIndex.length);
+    }
+
+    /**
+     * This can, potentially, be removed in the future by CASSANDRA-18597
+     */
+    private int recurseBalanceTree(int nodeID, int offset, int count, int treeDepth, byte[] splitValues, List<byte[]> leafBlockStartValues)
+    {
+        treeDepth++;
+        if (count == 1)
+        {
+            // Leaf index node
+            System.arraycopy(leafBlockStartValues.get(offset), 0, splitValues, nodeID * bytesPerValue, bytesPerValue);
+        }
+        else if (count > 1)
+        {
+            // Internal index node: binary partition of count
+            int countAtLevel = 1;
+            int totalCount = 0;
+            while (true)
+            {
+                int countLeft = count - totalCount;
+                if (countLeft <= countAtLevel)
+                {
+                    // This is the last level, possibly partially filled:
+                    int lastLeftCount = Math.min(countAtLevel / 2, countLeft);
+                    assert lastLeftCount >= 0;
+                    int leftHalf = (totalCount - 1) / 2 + lastLeftCount;
+
+                    int rootOffset = offset + leftHalf;
+
+                    System.arraycopy(leafBlockStartValues.get(rootOffset), 0, splitValues, nodeID * bytesPerValue, bytesPerValue);
+
+                    // TODO: we could optimize/specialize, when we know it's simply fully balanced binary tree
+                    // under here, to save this while loop on each recursion
+
+                    // Recurse left
+                    int leftTreeDepth = recurseBalanceTree(2 * nodeID, offset, leftHalf, treeDepth, splitValues, leafBlockStartValues);
+
+                    // Recurse right
+                    int rightTreeDepth = recurseBalanceTree(2 * nodeID + 1, rootOffset + 1, count - leftHalf - 1, treeDepth, splitValues, leafBlockStartValues);
+                    return Math.max(leftTreeDepth, rightTreeDepth);
+                }
+                totalCount += countAtLevel;
+                countAtLevel *= 2;
+            }
+        }
+        else
+        {
+            assert count == 0;
```

Comment:
Is it illegal for us to get here now?

It seems like if you had `count == 0` the returned depth would be 2?

Reply:
Answering this and below here.

I have refined this calculation to be more accurate and have added testing to test that the tree depth is correct. I have moved the increment to only increment for tree nodes and leaf nodes. The tree depth starts at one to account for the root node.